monitors/ContractMonitor.js

```diff
@@ -0,0 +1,173 @@
+const { Logger } = require("../financial-templates-lib/logger/Logger");
+
+const { createFormatFunction, createEtherscanLinkMarkdown } = require("../common/FormattingUtils");
+const networkUtils = require("../common/PublicNetworks");
+
+class ContractMonitor {
+  constructor(expiringMultiPartyEventClient, account, umaLiquidatorAddress, umaDisputerAddress) {
+    // Bot and ecosystem accounts.
+    this.account = account;
+    this.umaLiquidatorAddress = umaLiquidatorAddress;
+    this.umaDisputerAddress = umaDisputerAddress;
+
+    // Previous contract state used to check for new entries between calls
+    this.lastLiquidationBlockNumber = 0;
+    this.lastDisputeBlockNumber = 0;
+    this.lastDisputeSettlementBlockNumber = 0;
+
+    // EMP event client to read latest contract events
+    this.empEventClient = expiringMultiPartyEventClient;
+    this.web3 = this.empEventClient.web3;
+
+    // Contract constants
+    //TODO: replace this with an actual query to the collateral currency symbol
+    this.collateralCurrencySymbol = "DAI";
+    this.syntheticCurrencySymbol = "UMATEST";
+
+    // TODO: get the decimals of the collateral currency and use this to scale the output appropriately for non 1e18 colat
+    this.formatDecimalString = createFormatFunction(this.web3, 2);
+  }
+
+  // Calculate the collateralization Ratio from the collateral, token amount and token price
+  // This is cr = [collateral / (tokensOutstanding * price)] * 100
+  calculatePositionCRPercent = (collateral, tokensOutstanding, priceFunction) => {
+    return this.web3.utils
+      .toBN(collateral)
+      .mul(this.web3.utils.toBN(this.web3.utils.toWei("1")))
+      .mul(this.web3.utils.toBN(this.web3.utils.toWei("1")))
+      .div(this.web3.utils.toBN(tokensOutstanding).mul(this.web3.utils.toBN(priceFunction.toString())))
+      .muln(100);
```

Comment:
Two comments:
1. Does this need to replicate how we compute it in the contract? This way means we have less rounding, but it doesn't mirror the contract's computation. I'm totally on board as long as we're okay having a more accurate read here than the contract does. This could mean that a liquidation looks invalid from the bot's perspective, but the contract thinks it's valid (although these cases would probably be rare).
2. I think you're correct in assuming that `BN`s don't overflow/have a max size, but we may want to double check.
3. Isn't the price function supposed to be a function? It seems like you're treating it as a value here.

Reply:
1. the value returned here does nothing more than inform the console log, something like this would be sent in slack.
```
...Sponsor collateralization was 150.01%
```
So the precision or rounding has no meaningful effect as this output is truncated to two decimal points. Additionally this calculation is just to visually show in the slack bot what the CR was at liquidation time so extra precision is neither here nor their in this case.

This is however a good point in general. We currently do this in the `ExpiringMultiPartyClient` [here](https://github.com/UMAprotocol/protocol/blob/5711c55d75db38651c65b9b78ae134c66ea72502/financial-templates-lib/ExpiringMultiPartyClient.js#L117) and your point about off-chain infrastructure representing different values to the contracts is more relevant their as this could mean an invalid liquidation attempt in a very strange edge case.

2. I'm pretty sure that this is not an issue. I looked at the bignumber documentation and it says:
```
The largest possible magnitude of a finite BigNumber is 9.999...e+1000000000.
The smallest possible magnitude of a non-zero BigNumber is 1e-1000000000.
```
so this should not pose a problem. ALso we'd have issues in other places in the stack.

3. It is. In this PR it's a string. I followed the pattern used in the liquidator bot and which was changed between PRs to correctly use a function over a string. I have refactored this accordingly.